File: ayeon/괄호변환.cpp
/* PROCEEDING */
==============================================================

string solution(string w)
	if ( w == "" ) return w;
	
	
	same = 1;
	different=0;
	while(same != different)
	{
		if(w[0] == w[i++]) same++;
		else different++;
	}
	u <- slice w [ 0 to i ]
	v <- slice w [ i to w.size()-1 ] 

	if (check (u))
		return u + solution(v);
	else
	{
		answer = "(" + solution(v) + ")"
		u <- delete u[0], u[u.size()-1]
		answer = answer + reverse(u);
		return answer;
	}



// function for check(string w)

	i=0
	stack.push(w[i++])
	while(!stack.empty())
	{
		if(stack.top() == w[i]) stack.push(w[i])
		else stack.pop();
		i++;
	}
	u = slice w[0] ~ w[i] 
	v = w[i+1] ~ w[]

File: minnkyoung/괄호변환.py
def solution(p):
    if not p:
        return ""
    u,v = seperate(p)
    
    if isBalanced(u):
        return u + solution(v) 

    else:
        answer = '('
        answer += solution(v)
        answer += ')'

        for i in u[1:-1]:
            if i == '(':
                answer += ')'
            else:
                answer += '('
                
        return answer

def seperate(p):
    left = 0
    right = 0
    for i in range(len(p)):
        if p[i] == '(':
            left += 1
        elif p[i] == ')':
            right += 1
        if left == right:
            return p[:i+1], p[i+1:] 

def isBalanced(u):
    stack = []

    for i in u:
        if i == '(':
            stack.append(i)
        else:
            if not stack:
                return False
            stack.pop()
    return True